driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -236,6 +236,11 @@ public IntegerExpression parseInteger() {
         return new MqlExpression<>(ast("$toLong"));
     }
 
+    @Override
+    public NumberExpression parseNumber() {
+        return new MqlExpression<>(ast("$toDecimal"));
```

Comment:
This seems like a problem.  While it's true that Decimal128 is the most "inclusive" numeric type, applications will likely be surprised to see a Decimal128 in a projection and it will lead to runtime errors.  

Is there a tension between the BSON type system, which distinguishes between the different numeric types, and this API, which generally does not?

Reply:
Yes, but the tension is between the BSON type system and the MQL type system, which are both exposed by the server. This API conforms to the MQL type system. (This topic has also come up [here](https://github.com/mongodb/mongo-java-driver/pull/1052#discussion_r1032648066).)

The MQL type system seems to have arisen organically, so there is no proper specification. In practice, there is no distinction between decimals and doubles, or longs and ints. This is similar to how the Java `List` API is insensitive to any difference between Linked and ArrayLists.

The expressions API presently does not interact with either the BSON type system (or even the Java type system, except in the conversion methods on `Expressions` and on the helper methods).
https://www.mongodb.com/docs/manual/reference/operator/aggregation/isNumber/
Of the places where a BSON type is promoted (division, overflow...), this is one of

Reply:
Yes, but the tension is between the BSON type system and the MQL type system, which are both exposed by the server. This API conforms to the MQL type system. (This topic has also come up [here](https://github.com/mongodb/mongo-java-driver/pull/1052#discussion_r1032648066).)

The MQL type system seems to have arisen organically, so there is no proper specification. In practice, it makes no distinction between decimals and doubles, or longs and ints. This is similar to how the Java `List` API is insensitive to any difference between Linked and ArrayLists.

Presently, the expressions API does not interact with the BSON type system (or even the Java type system, except in the conversion methods on `Expressions` and on the helper methods).

I agree that we should make an effort to round-trip underlying types (though there are cases where the existing MQL API makes this impossible). There are several ways to proceed:

1. Break the encapsulation of BSON in this API. This mixes the type systems, and is confusing to users. Users would have to explicitly manage types from a different abstraction layer when they are already writing difficult functional code.
2. Handle unusual cases like this in the driver (including the more-subtle `1/2` division and overflow promotions). This will not be possible to handle all cases.
3. Have the server provide a `parseNumber` function to go with [isNumber](https://www.mongodb.com/docs/manual/reference/operator/aggregation/isNumber/). I see no way to reasonably emulate this.
4. Do nothing. Possibly document that parseNumber returns a decimal (though this is a bit like documenting that the Java streams `collect` method returns an ArrayList).

I think 4 is our best option at present, all things considered.

Reply:
I considered some other options, and I think what I pushed up in the latest commit should resolve the problem. `toNumber` will parse to decimal by default, and `toInteger` to int32 by default, falling back to the broader type when the number is too large.

As an aside, I found out that division turns all integer numbers into doubles, even in cases where the division could yield a simple integers, like `2/1`. I added tests.

Reply:
I like the idea, but it still doesn't seem sufficient for all practical needs.  One of the different things about decimal 128 is that precision is preserved even in the face of trailing zeros (BigDecimal is the same).  So these strings are all different decimal 128 values: "2", "2.0", "2.00", and so on.  So there are times when those values exist as strings and an application needs to convert them to the correct decimal 128 value.  But if i understand correctly, this PR will convert them all to the same double value.

Unfortunately I don't see a way to avoid making this API aware of the different number BSON types.

Reply:
This is interesting. All of the following pass.
```
assertExpression(
        true,
        of(Decimal128.parse("2")).eq(of(Decimal128.parse("2.00"))),
        "{'$eq': [{'$numberDecimal': '2'}, {'$numberDecimal': '2.00'}]}");

assertExpression(
        Decimal128.parse("4.00"),
        of("2.00").parseNumber().multiply(of(2)),
        "{'$multiply': [{'$toDecimal': '2.00'}, 2]} ");
assertExpression(
        Decimal128.parse("4"),
        of("2").parseNumber().multiply(of(2)),
        "{'$multiply': [{'$toDecimal': '2'}, 2]} ");

assertExpression(
        "2.00000",
        of("2").parseNumber().round(of(5)).asString());
```
I see no way to check precision, or to otherwise reasonably tell decimal values apart using the MQL API.

I would agree that we need an extra method for place-sensitive parsing. This looks comparable to what we are doing for dates, where we have overrides for cases where the date string has a format that we must be sensitive to (except if the dates were actually timezoned dates, but with no way to check the timezone without writing to a string). I'll think about this.

Reply:
Removed `toNumber`, as discussed.